packages/common/core-utils/src/compare.ts

```diff
@@ -25,9 +27,23 @@ export const compareArrays = <T>(
 	//       '===' and 'Object.is()' tied.
 	//       Trivial acceptance adds no measurable overhead.
 	//       30% penalty vs. baseline for exported function [node 14 x64].
-	return (
-		left === right || // Trivial acceptance: 'left' and 'right' are the same instance
-		(left.length === right.length && // Trivial rejection: 'left' and 'right' are different lengths
-			left.every((leftItem, index) => comparator(leftItem, right[index], index)))
-	);
+	if (left === right) {
+		// Trivial acceptance: 'left' and 'right' are the same instance
+		return true;
+	}
+
+	if (left.length !== right.length) {
+		// Trivial rejection: 'left' and 'right' are different lengths
+		return false;
+	}
+
+	for (const [index, leftItem] of left.entries()) {
+		const rightItem = right[index];
+		assert(rightItem !== undefined, "rightItem is undefined in compareArrays");
```

Comment:
This executes at runtime which I think hurts the performance intent of this function.
Can you get build success using an index based for loop?

Reply:
Also, you should mark this as draft since not in a state you want to commit.

Reply:
Because of the perf sensitivity, and because in this case we happen to know something that TypeScript doesn't, or doesn't remember. That is, we know that the lengths of the arrays is the same, and the arrays are typed `readonly` so we know there's no manipulation in the function itself (well we could know that from code inspection too but the typing enforces it).

My JS runtime knowledge is rusty and but I don't _think_ we have to worry about external modifications of the input arrays within the function. If I'm wrong, though, then discount everything I say. :)

Finally, the motivation behind adopting this setting is the ts-semver guidance from https://www.semver-ts.org/formal-spec/5-compiler-considerations.html#strictness, with the goal of protecting us against type-incompatible changes in our API. I don't think that really applies here.

With all this in mind, I think this may be an appropriate place to use a non-null assertion (`!`) with a long comment explaining why we know better than the compiler. (and again, if I am wrong about that then disregard)

Reply:
Yes, when there isn't a different way to write the code that allows the tooling to understand it is legitimate to bring our knowledge to bear. A non-null assertion should resolve the tsc complaint and then we would need to use a lint disable for the non-null or other type assertion which should have the appropriate comment.
And you are correct Tyler about execution. We are running under assumption of no-preemption environment. So, the only opportunity is for someone to write a crass `comparator` that actually manipulates either of the arrays. Well, in that case whatever.
However, in this case instead of a non-null assertion I would suggest using `as T`, the difference is that `T` is allowed to be `null`. It is very subtle.
So:
`right[index] as T` with something like
```
// right index access is protected by length test above, right[index] is not undefined.
// eslint-disable-next-line @typescript-eslint/no-unnecessary-type-assertion
```

Reply:
This works, should I keep it as this or is there a better way??
```ts
for (let index = 0; index < left.length; index++) {
  const leftItem = left[index] as T;
  const rightItem = right[index] as T;
  if (!comparator(leftItem, rightItem, index)) {
	  return false;
  }
}
```

Reply:
While as T is cool, it's still a type cast so be very careful when using it. :)